Add strict mode to AssemblyRegionWalkerSpark and HaplotypeCallerSpark

tomwhite opened the pull request:
There are tests for both the strict and non-strict modes for HaplotypeCallerSpark and for ExampleAssemblyRegionWalkerSpark. The strict modes produce identical results to the walker versions.

Strict mode is off by default and still needs work to scale to exome-sized data (another PR). This is an improvement over the current HaplotypeCallerSpark implementation since it fixes two bugs (reads overlapping more than two intervals; and editIntervals not being picked up) that caused the output to differ significantly from the walker version. While the output does still differ in non-strict mode, the difference is a lot less (compare expected.testVCFMode.gatk4.vcf and expected.testVCFMode.gatk4.nonstrict.vcf).

tomwhite commented on src/main/java/org/broadinstitute/hellbender/engine/spark/FindAssemblyRegionsSpark.java:
```diff
@@ -0,0 +1,215 @@
+        JavaRDD<AssemblyRegion> assemblyRegions = assemblyRegionShardedReads.map((Function<Shard<GATKRead>, AssemblyRegion>) shard -> toAssemblyRegion(shard, header));
+
+        // 7. Add reference and feature context.
+        return assemblyRegions.mapPartitions(getAssemblyRegionWalkerContextFunction(referenceFileName, bFeatureManager));
```
These 7 steps make up the salient part of the strict algorithm.

jamesemery commented on src/main/java/org/broadinstitute/hellbender/engine/spark/FindAssemblyRegionsSpark.java:
```diff
@@ -0,0 +1,215 @@
+        // TODO: interfaces could be improved to avoid casting
+        ReadlessAssemblyRegion readlessAssemblyRegion = (ReadlessAssemblyRegion) ((ShardBoundaryShard<GATKRead>) shard).getShardBoundary();
+        int extension = Math.max(shard.getInterval().getStart() - shard.getPaddedInterval().getStart(), shard.getPaddedInterval().getEnd() - shard.getInterval().getEnd());
+        AssemblyRegion assemblyRegion = new AssemblyRegion(shard.getInterval(), Collections.emptyList(), readlessAssemblyRegion.isActive(), extension, header);
```
I see the issue with this interface... It looks like you may be able to make regular assemblyRegion also extend ShardBoundaryShard. I'm not sure that helps anything really... but then at least AssemblyRegion and ReadlessAssemblyRegion are related in the class hirerarchy.

jamesemery replied:
that is though an abstract interface ofcourse

tomwhite replied:
Thanks for the suggestion. I need to spend a bit more time trying to improve this. (Could be done in another issue or PR.)

jamesemery commented on src/main/java/org/broadinstitute/hellbender/engine/spark/FindAssemblyRegionsSpark.java:
```diff
@@ -0,0 +1,215 @@
+
+        // 4. Pull the assembly region boundaries down to the driver, so we can fill in reads.
+        List<ShardBoundary> assemblyRegionBoundaries = readlessAssemblyRegions
+                .map((Function<ReadlessAssemblyRegion, ShardBoundary>) FindAssemblyRegionsSpark::toShardBoundary)
```
Why do you need toShardBoundary here when a ReaderlessAssemblyRegion is a shard bounadry?

tomwhite replied:
It's a problem with generics... `List<A>` cannot be substituted by `List<B>` even if `B` is a subclass of `A`. There should be a way of improving the code in this case though.

tomwhite replied:
I fixed this by generalizing the type in `SparkSharder`.

jamesemery commented on src/main/java/org/broadinstitute/hellbender/engine/spark/FindAssemblyRegionsSpark.java:
```diff
@@ -0,0 +1,215 @@
+
+            return Utils.stream(shardedReadIterator)
+                    .map(shardedRead -> new ShardToMultiIntervalShardAdapter<>(shardedRead))
+                    // TODO: reinstate downsampling (not yet working)
```
We should think about how we want to do this. Notably however, as of this implementation you need to add the same downsampling to the reads at step 5, as this is currently only hooked up to downsample for the band pass filtering but not on the reads when they get re-added to the filter. An obvious solution might involve persisiting the underlying bam but obviously thats expensive, the other is that we make the random seed in some way caclulatable, like base it on a hash of the read if necessary, that way we could get an exact mach with the regular haplotype caller.

tomwhite replied:
The reason that I excluded this is because I found an underlying bug in the downsampling code where it would not do the downsampling correctly unless the iterator was created twice. (This needs further investigation.)

But you are right that the reads downsampling needs to have been carried out for the reads too, which isn't currently happening in this code. Let's explore your ideas further.

jamesemery commented on src/main/java/org/broadinstitute/hellbender/engine/ReadlessAssemblyRegion.java:
```diff
@@ -0,0 +1,58 @@
+ * A cut-down version of {@link AssemblyRegion} that doesn't store reads, used in the strict implementation of
+ * {@link org.broadinstitute.hellbender.engine.spark.FindAssemblyRegionsSpark}.
+ */
+public class ReadlessAssemblyRegion extends ShardBoundary {
```
There should probably be a common interface linking this to AssemblyRegion, or a different name distinguishing it.

tomwhite replied:
Agreed.

tomwhite commented on src/main/java/org/broadinstitute/hellbender/engine/ActivityProfileStateIterator.java:
```diff
@@ -0,0 +1,92 @@
+
+        // We wrap our LocusIteratorByState inside an IntervalAlignmentContextIterator so that we get empty loci
+        // for uncovered locations. This is critical for reproducing GATK 3.x behavior!
+        LocusIteratorByState libs = new LocusIteratorByState(readShard.iterator(), DownsamplingMethod.NONE, false, ReadUtils.getSamplesFromHeader(readHeader), readHeader, includeReadsWithDeletionsInIsActivePileups);
```
`ActivityProfileStateIterator` is the part of `AssemblyRegionIterator` that just does the iteration over activity profiles. It's the same code and should arguably be refactored to share code (e.g. `AssemblyRegionIterator` might compose `ActivityProfileStateIterator` and `AssemblyRegionFromActivityProfileStateIterator`, although it's not that simple, since `AssemblyRegionIterator` does read caching, which we don't want in the Spark case.

Also, I don't think we see downsampling kicking in for the tests, so it's probably not the reason for difference in the tests. That seems to be read shard boundary artifacts.

tomwhite commented on src/main/java/org/broadinstitute/hellbender/engine/spark/FindAssemblyRegionsSpark.java:
```diff
@@ -0,0 +1,215 @@
+import java.util.List;
+import java.util.function.Supplier;
+
+public class FindAssemblyRegionsSpark {
```
Done

tomwhite commented on src/test/java/org/broadinstitute/hellbender/engine/spark/SparkSharderUnitTest.java:
```diff
@@ -114,6 +114,66 @@ public void testLocatablesPerShard() throws IOException {
+                new TestRead(17, 19), new TestRead(21, 23), new TestRead(25, 27)
+        );
+
+        List<SimpleInterval> intervals = ImmutableList.of(
```
This is testing the overlaps part of `SparkSharder.locatablesPerShard`, padding is added by the `shard` method, which is exercise by other tests.

tomwhite commented on src/test/java/org/broadinstitute/hellbender/engine/spark/SparkSharderUnitTest.java:
```diff
@@ -114,6 +114,66 @@ public void testLocatablesPerShard() throws IOException {
         assertEquals(next._2(), ImmutableList.of());
     }
 
+    @Test
```
Agreed - already tested by `testSingleContig` and `testContigBoundary`, but even more tests would be good.

tomwhite commented on src/main/java/org/broadinstitute/hellbender/engine/AssemblyRegionFromActivityProfileStateIterator.java:
```diff
@@ -0,0 +1,141 @@
+    private final int minRegionSize;
+    private final int maxRegionSize;
+    private final int assemblyRegionPadding;
+    private final double activeProbThreshold;
```
Removed

tomwhite commented on src/main/java/org/broadinstitute/hellbender/engine/AssemblyRegionFromActivityProfileStateIterator.java:
```diff
@@ -0,0 +1,141 @@
+/**
+ * Given an iterator of {@link ActivityProfileState}, finds {@link AssemblyRegion}s.
+ *
+ * This iterator and {@link ActivityProfileStateIterator} represent the core of the {@link AssemblyRegionWalker} traversal.
```
Fixed

tomwhite commented on src/main/java/org/broadinstitute/hellbender/engine/AssemblyRegionFromActivityProfileStateIterator.java:
```diff
@@ -0,0 +1,141 @@
+    private AssemblyRegion loadNextAssemblyRegion() {
+        AssemblyRegion nextRegion = null;
+
+        while ( activityProfileStateIterator.hasNext() && nextRegion == null ) {
```
I think so. Note that this is existing code from the walker version.

tomwhite commented on src/main/java/org/broadinstitute/hellbender/engine/AssemblyRegionFromActivityProfileStateIterator.java:
```diff
@@ -0,0 +1,141 @@
+
+        this.readyRegion = null;
+        this.pendingRegions = new ArrayDeque<>();
+        this.activityProfile = new BandPassActivityProfile(null, maxProbPropagationDistance, activeProbThreshold, BandPassActivityProfile.MAX_FILTER_SIZE, BandPassActivityProfile.DEFAULT_SIGMA, readHeader);
```
Existing walker code. I think https://github.com/broadinstitute/gatk/issues/5387 is related.

tomwhite commented on src/main/java/org/broadinstitute/hellbender/tools/HaplotypeCallerSpark.java:
```diff
@@ -153,13 +138,38 @@ private static void processAssemblyRegions(
+        try {
+            super.runTool(ctx);
+        } catch (Exception e) {
+            if (e.getCause() instanceof UserException) {
```
Otherwise it is propagated as a SparkException, which is not what the user is expecting.

tomwhite commented on src/main/java/org/broadinstitute/hellbender/tools/spark/pipelines/ReadsPipelineSpark.java:
```diff
@@ -138,6 +138,9 @@
     @ArgumentCollection
     public HaplotypeCallerArgumentCollection hcArgs = new HaplotypeCallerArgumentCollection();
 
+    @Argument(doc = "whether to use the strict implementation or not (defaults to the faster implementation that doesn't strictly match the walker version)", shortName = "strict", fullName = "strict", optional = true)
```
Good to know that about Barclay - fixed. The argument is inherited from `AssemblyRegionWalkerSpark` by `HaplotypeCallerSpark`.

tomwhite commented on src/main/java/org/broadinstitute/hellbender/utils/activityprofile/ActivityProfileStateRange.java:
```diff
@@ -0,0 +1,79 @@
+import java.util.List;
+
+/**
+ * An efficient way of representing a set of {@link ActivityProfileState}s in an interval; used by Spark.
```
In an earlier version that didn't have `ActivityProfileStateRange` (i.e. was just using `ActivityProfileState` for each position) I would see OOME when running on a full contig (even chr21). So this is a necessary start, but there's likely more to do on reducing the serialized size of objects.

tomwhite commented on src/main/java/org/broadinstitute/hellbender/engine/spark/FindAssemblyRegionsSpark.java:
```diff
@@ -0,0 +1,215 @@
+                    .map(shardedRead -> new ShardToMultiIntervalShardAdapter<>(
+                            new DownsampleableSparkReadShard(
+                                    new ShardBoundary(shardedRead.getInterval(), shardedRead.getPaddedInterval()), shardedRead, readsDownsampler)))
+                    .map(shardedRead -> {
```
Done

tomwhite commented on src/main/java/org/broadinstitute/hellbender/engine/spark/FindAssemblyRegionsSpark.java:
```diff
@@ -0,0 +1,215 @@
+        // produce assembly regions that are identical to those produced by AssemblyRegionWalker.
+        // This step requires a shuffle, but the amount of data in the ActivityProfileStateRange should be small, so it
+        // should not be prohibitive.
+        JavaPairRDD<String, Iterable<ActivityProfileStateRange>> contigToGroupedStates = activityProfileStates
```
I agree. Where is the code that does this BTW?

jamesemery replied:
@tomwhite That happens in this wdl https://github.com/broadinstitute/dsde-pipelines/blob/develop/genomes_in_the_cloud/single_sample/PairedSingleSampleWf.wdl. Looking at the json they seem to have the file storeed here: gs://broad-references/hg38/v0/wgs_calling_regions.hg38.interval_list

tomwhite commented on src/test/java/org/broadinstitute/hellbender/tools/HaplotypeCallerSparkIntegrationTest.java:
```diff
@@ -25,9 +27,88 @@
+     * Make sure that someone didn't leave the UPDATE_EXACT_MATCH_EXPECTED_OUTPUTS toggle turned on
+     */
+    @Test
+    public void assertThatExpectedOutputUpdateToggleIsDisabled() {
```
`testVCFModeIsConsistentWithPastResults` tests Spark strict mode is identical to the walker version. `testNonStrictVCFModeIsConsistentWithPastResults` test Spark non-strict mode doesn't regress. The other (existing) tests test concordance with GATK3 and 4 (walkers). Is there another test you'd like to add?

It's not a part of this PR, but I'd like to be able to run `HaplotypeCallerIntegrationTest` against the Spark tool.

jamesemery replied:
If `testVCFModeIsConsistentWithPastResults` is indeed identical could we factor out a common haplotype caller integration tests class in order to enforce that (similar to what we used to have for the markduplicates tests)

tomwhite commented:
Thanks for the review @jamesemery. I've addressed your comments. A few outstanding issues:
* `ActivityProfileStateIterator` and `AssemblyRegionFromActivityProfileStateIterator` duplicate parts of `AssemblyRegionIterator`, so it would be nice to remove the code duplication. Not totally straightforward as the latter does read caching, but the first two don't (Spark shouldn't be caching reads).
* Downsampling needs more work. I would be OK doing that separately, since I've only ever seen it when running on a full genome, and the new strict code needs more work to work on a full genome (I've only got it running on an exome so far).
* There are some improvements we could make to `ReadlessAssemblyRegion` regarding Java interface design and generics, but I'm not sure what they are yet.

I'm not sure if these are blockers, since the strict codepath is a new option (off by default), but would like to know what you and @jonn-smith think.

tomwhite commented on src/main/java/org/broadinstitute/hellbender/engine/spark/FindAssemblyRegionsSpark.java:
```diff
@@ -0,0 +1,216 @@
+        };
+    }
+
+    public static JavaRDD<AssemblyRegionWalkerContext> getAssemblyRegionsStrict(
```
Done

tomwhite commented on src/main/java/org/broadinstitute/hellbender/engine/spark/FindAssemblyRegionsSpark.java:
```diff
@@ -0,0 +1,216 @@
+            final AssemblyRegionArgumentCollection assemblyRegionArgs,
+            final boolean includeReadsWithDeletionsInIsActivePileups) {
+        return (FlatMapFunction<Iterator<Shard<GATKRead>>, ActivityProfileStateRange>) shardedReadIterator -> {
+            ReferenceDataSource reference = referenceFileName == null ? null : new ReferenceFileSource(IOUtils.getPath(SparkFiles.get(referenceFileName)));
```
Done

tomwhite commented on src/main/java/org/broadinstitute/hellbender/engine/spark/FindAssemblyRegionsSpark.java:
```diff
@@ -0,0 +1,216 @@
+        return (FlatMapFunction<Iterator<Shard<GATKRead>>, ActivityProfileStateRange>) shardedReadIterator -> {
+            ReferenceDataSource reference = referenceFileName == null ? null : new ReferenceFileSource(IOUtils.getPath(SparkFiles.get(referenceFileName)));
+            final FeatureManager features = bFeatureManager == null ? null : bFeatureManager.getValue();
+            AssemblyRegionEvaluator assemblyRegionEvaluator = supplierBroadcast.getValue().get(); // one AssemblyRegionEvaluator instance per Spark partition
```
Done

tomwhite commented on src/main/java/org/broadinstitute/hellbender/engine/spark/FindAssemblyRegionsSpark.java:
```diff
@@ -0,0 +1,216 @@
+ */
+public class FindAssemblyRegionsSpark {
+
+    public static JavaRDD<AssemblyRegionWalkerContext> getAssemblyRegionsFast(
```
Done

tomwhite commented on src/main/java/org/broadinstitute/hellbender/engine/spark/FindAssemblyRegionsSpark.java:
```diff
@@ -0,0 +1,216 @@
+            final Broadcast<FeatureManager> bFeatureManager) {
+
+        return (FlatMapFunction<Iterator<AssemblyRegion>, AssemblyRegionWalkerContext>) assemblyRegionIter -> {
+            ReferenceDataSource reference = referenceFileName == null ? null : new ReferenceFileSource(IOUtils.getPath(SparkFiles.get(referenceFileName)));
```
Done

tomwhite commented:
Thanks for the review @jonn-smith. I've addressed all your comments, and those of @jamesemery. Please approve so we can merge.

tomwhite commented:
Thanks @jamesemery. I've successfully run with an exome in #5475; I'll try with a genome separately.
